core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java

```diff
@@ -0,0 +1,72 @@
+package org.neo4j.graphalgo.core.utils;
+
+/**
+ * A structure to manage disjoint sets of ints e.g. nodeIds
+ *
+ * @author mknblch
+ */
+public class DisjointSetStruct {
+
+    private final int[] parent;
+    private final int[] depth;
+
+    /**
+     *
+     * @param capacity the capacity (maximum node id)
+     */
+    public DisjointSetStruct(int capacity) {
+        parent = new int[capacity];
+        depth = new int[capacity];
+        for (int i = 0; i < capacity; i++) {
+            parent[i] = i;
+        }
+    }
+
+    /**
+     * find setId of element p.
+     * @param p the element in the set we are looking for
+     * @return an id of the set it belongs to
+     */
+    public int find(int p) {
+        if (parent[p] == p) return p;
+        // path compression optimization
+        parent[p] = find(parent[p]); // balance tree while traversing
```

Comment:
recursion, can we have huge depths in larger graphs? perhaps replace with iteration? (i.e. a loop)